### PR TITLE
Institutions 2.1

### PR DESCRIPTION
**Pull request by hmoco**

### CODE REVIEW

File: scripts/populate_institutions.py [hunk 1]
```diff
@@ -55,7 +55,28 @@ def main(env):
...
+            'domain': ['staging-osf.cos.io'],
+            'description': None,
+            'email_domain': ['cos.io'],
+            'banner_name': 'cos-banner.png'
+        },
+        {
+            'name': 'University of Southern California',
+            '_id': 'usc',
+            'logo_name': 'usc-shield.jpg',
+            'auth_url': 'MICHAEL CHANGE THIS',
```
Comment (sloria) [inline scripts/populate_institutions.py]: ping @icereval

File: website/static/js/institutionSignIn.js [hunk 2]
```diff
@@ -18,11 +18,26 @@ var ViewModel = function() {
...
+            self.instNames(
+                response.data.map(function(item){
+                    var name = item.attributes.name;
+                    self.insts[name] = item.attributes.auth_url + '&target=' + encodeURIComponent(window.contextVars.institution_redirect);
+                    self.insts[item.id] = name;
+                    if (item.attributes.auth_url){
+                        return name;
+                    }
+                }).filter(function (item) {
+                   return typeof item !== 'undefined';
```
Comment (sloria) [inline website/static/js/institutionSignIn.js]: This is suspect; why would the response contain `undefined`?

File: website/static/js/institutionSignIn.js [hunk 3]
```diff
@@ -18,11 +18,26 @@ var ViewModel = function() {
...
+                })
+            );
+            var inst_redirect = decodeURIComponent(decodeURIComponent(window.contextVars.institution_redirect));
```
Comment (sloria) [inline website/static/js/institutionSignIn.js]: As per COSDev, use `lowerCamelCase` for variable names.

File: website/static/js/institutionSignIn.js [hunk 4]
```diff
@@ -18,11 +18,26 @@ var ViewModel = function() {
...
+            if (inst_redirect){
+                var inst_id = inst_redirect.split('institutions')[1];
+                inst_id = inst_id ? inst_id.split('/')[1] : false;
```
Comment (sloria) [inline website/static/js/institutionSignIn.js]: `lowerCamelCase` here, too

File: website/templates/institution.mako [hunk 5]
```diff
@@ -8,17 +8,21 @@
     <div class="alert alert-info"><strong>NOTICE: </strong>Forks, registrations, and uploads will be temporarily disabled while the OSF undergoes a hardware upgrade. These features will return shortly. Thank you for your patience.</div>
 % endif
     <div id="inst">
-<div class="dashboard-header">
+<div class="dashboard-header" style="background: white">
```
Comment (sloria) [inline website/templates/institution.mako]: Avoid inline styling. How about adding a new class to `dashboard-page.css`:

```css
.dashboard-header.dashboard-header-institution {
    background: none;
}
```

Then applying

```html
class="dashboard-header dashboard-header-institution"
```

File: website/static/js/institutionSignIn.js (hunk 2 above)
Comment (sloria) [inline website/static/js/institutionSignIn.js]: Oh, I see now; it's possible that the `map` function will return `undefined`. How about doing the filter first?

```js
response.data.filter(function(item) { return item.attributes.auth_url;}).map(function(item) {
    // ...
})
```

Comment (hmoco) [inline website/static/js/institutionSignIn.js]: Response doesn't. Not returning anything on ` .map ` (which happens if the item does not have an ` auth_url `) adds an undefined object to the list. This is here to clean up those undefineds.

Comment (hmoco) [inline website/static/js/institutionSignIn.js]: Hah, we commented at the same time. Yep that change seems reasonable.

Comment (sloria) [inline website/static/js/institutionSignIn.js]: Updated above comment with more detail.

File: website/static/js/institutionSignIn.js [hunk 6]
```diff
@@ -18,11 +18,26 @@ var ViewModel = function() {
...
         ).done(function (response) {
-            self.instNames(response.data.map(function(item){
-                var name = item.attributes.name;
-                self.insts[name] = item.attributes.auth_url + '&target=' + encodeURIComponent(window.contextVars.institution_redirect);
-                return name;
-            }));
+            self.instNames(
+                response.data.map(function(item){
+                    var name = item.attributes.name;
+                    self.insts[name] = item.attributes.auth_url + '&target=' + encodeURIComponent(window.contextVars.institution_redirect);
```
Comment (sloria) [inline website/static/js/institutionSignIn.js]: What if `auth_url` isn't defined?

File: website/project/views/node.py [hunk 7]
```diff
@@ -249,7 +249,8 @@ def node_setting(auth, node, **kwargs):
 
     #check institutions:
     try:
-        inst = Institution.find_one(Q('email_domain', 'eq', auth.user.username.split('@')[1]))
+        email_domains = [email.split('@')[1] for email in auth.user.emails]
+        inst = Institution.find_one(Q('email_domain', 'in', email_domains))
```
Comment (sloria) [inline website/project/views/node.py]: This looks incorrect, given that `email_domain` is a list field: https://github.com/CenterForOpenScience/osf.io/blob/54418bc5fbce98fdb892267a27a93a560b06433a/website/project/model.py#L3384 .

File: website/views.py [hunk 8]
```diff
@@ -79,15 +80,13 @@ def index():
     try:
         #TODO : make this way more robust
         inst = Institution.find_one(Q('domain', 'eq', request.host))
```
Comment (sloria) [inline website/views.py]: Do we store `domain` lower-cased? If so, you'll need to do `request.host.lower()`

Comment (sloria) [inline website/views.py]: This led me to find that `domain` is a [list field](https://github.com/CenterForOpenScience/osf.io/blob/54418bc5fbce98fdb892267a27a93a560b06433a/website/project/model.py#L3381). Therefore, it should probably be renamed to `domains`.

File: website/project/views/node.py (hunk 7 above)
Comment (hmoco) [inline website/project/views/node.py]: Apparently this works. Have done (minor) testing, and Seto said he has mongo docs that support it.

**Comment by sloria** [comment]
Pass finished.

**Comment by hmoco** [comment]
@sloria have responded to comments.

File: website/project/views/node.py (hunk 7 above)
Comment (sloria) [inline website/project/views/node.py]: OK. As suggested in a previous comment, `email_domain` should probably be renamed to `email_domains` to make it clear that it's a list.

File: website/static/js/institutionSignIn.js [hunk 9]
```diff
@@ -18,11 +18,28 @@ var ViewModel = function() {
...
-            }));
+            self.instNames(
+                response.data.filter(function (item) {
+                    return item.attributes.auth_url;
+                }).map(function(item){
+                    return item.attributes.name;
+                })
+            );
+            response.data.forEach(function(item){
+                if (item.attributes.auth_url){
```
Comment (sloria) [inline website/static/js/institutionSignIn.js]: Minor: You're already filtering for data that has `auth_url` a few lines up; why not save that to a variable and re-use it?

```js
var validInstitutions = response.data.filter(...);
self.instNames(...map...)
validInstitutions.forEach(...)
```

File: website/project/model.py [hunk 10]
```diff
@@ -3381,7 +3381,8 @@ def require_approval(self, user, notify_initiator_on_complete=False):
     institution_domain = fields.StringField(list=True)
```
Comment (sloria) [inline website/project/model.py]: As suggested in a previous comment, this should probably be changed to `institution_domains`

**Comment by sloria** [comment]
Pass finished. Looks like https://github.com/CenterForOpenScience/osf.io/pull/5201/files#r56419747 is the only thing to be addressed.

**Pull request merged**